File: src/main/java/com/smarterd/domain/project/entity/Project.java
package com.smarterd.domain.project.entity;

import com.smarterd.domain.common.entity.BaseAuditEntity;
import com.smarterd.domain.common.exception.BusinessException;
import com.smarterd.domain.common.message.MessageCode;
import com.smarterd.domain.team.entity.Team;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.time.LocalDate;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.lang.Nullable;

/**
 * 프로젝트 엔티티.
 *
 * <p>팀({@link Team}) 소속으로 하나 이상의 {@link com.smarterd.domain.diagram.entity.Diagram}을 포함한다.
 * ERD 설계의 최상위 그룹 단위이다.</p>
 */
@Entity
@Table(name = "projects")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Project extends BaseAuditEntity {

    /** 프로젝트 고유 식별자 (자동 증가) */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** 프로젝트 이름 (최대 100자) */
    @Column(nullable = false, length = 100)
    private String name;

    /** 프로젝트 설명 (최대 500자, nullable) */
    @Column(length = 500)
    private String description;

    /** 발주사 (최대 200자, nullable) */
    @Nullable
    @Column(name = "client_company", length = 200)
    private String clientCompany;

    /** 수주사 (최대 200자, nullable) */
    @Nullable
    @Column(name = "contractor_company", length = 200)
    private String contractorCompany;

    /** 계약 금액 (원 단위, nullable) */
    @Nullable
    @Column(name = "contract_amount")
    private Long contractAmount;

    /** 프로젝트 시작일 (DATE, nullable) */
    @Nullable
    @Column(name = "project_start_date")
    private LocalDate projectStartDate;

    /** 프로젝트 종료일 (DATE, nullable) */
    @Nullable
    @Column(name = "project_end_date")
    private LocalDate projectEndDate;

    /** 사업 범위 (서술형, nullable) */
    @Nullable
    @Column(name = "project_scope", columnDefinition = "TEXT")
    private String projectScope;

    /** 소속 팀 */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = false)
    private Team team;

    /**
     * 프로젝트 엔티티를 생성한다.
     *
     * @param name        프로젝트 이름
     * @param description 프로젝트 설명 (nullable)
     * @param team        소속 팀
     */
    @Builder
    public Project(String name, String description, Team team) {
        this.name = name;
        this.description = description;
        this.team = team;
    }

    /**
     * 프로젝트 정보를 변경한다.
     *
     * @param name        새 프로젝트 이름
     * @param description 새 프로젝트 설명 (nullable)
     */
    public void update(String name, String description) {
        this.name = name;
        this.description = description;
    }

    /**
     * 사업 개요 정보를 갱신한다.
     *
     * @param clientCompany     발주사
     * @param contractorCompany 수주사
     * @param contractAmount    계약 금액
     * @param startDate         프로젝트 시작일
     * @param endDate           프로젝트 종료일
     * @param projectScope      사업 범위
     */
    public void updateBusinessOverview(
        @Nullable String clientCompany,
        @Nullable String contractorCompany,
        @Nullable Long contractAmount,
        @Nullable LocalDate startDate,
        @Nullable LocalDate endDate,
        @Nullable String projectScope
    ) {
        if (startDate != null && endDate != null && startDate.isAfter(endDate)) {
            throw new BusinessException(MessageCode.ERROR_BUSINESS_INVALID_PROJECT_PERIOD.code());
        }

        this.clientCompany = clientCompany;
        this.contractorCompany = contractorCompany;
        this.contractAmount = contractAmount;
        this.projectStartDate = startDate;
        this.projectEndDate = endDate;
        this.projectScope = projectScope;
    }
}

File: client/src/types/project.ts
/** 프로젝트 정보. 팀에 속하며 다이어그램을 포함한다. */
export interface Project {
  /** 프로젝트 고유 ID */
  id: number;
  /** 프로젝트 이름 */
  name: string;
  /** 프로젝트 설명 (nullable) */
  description: string | null;
  /** 소속 팀 ID */
  teamId: number;
  /** 프로젝트 생성 일시 (ISO 8601) */
  createdAt: string;
}

/** 사업 개요 조회 응답. BE 응답 11개 필드를 1:1로 매핑한다. */
export interface BusinessOverviewResponse {
  /** 프로젝트 ID */
  projectId: number;
  /** 프로젝트명 */
  projectName: string;
  /** 발주사명 */
  clientCompany: string | null;
  /** 수주사명 */
  contractorCompany: string | null;
  /** 계약 금액 (원화 단위 정수) */
  contractAmount: number | null;
  /** 프로젝트 시작일 (ISO 8601) */
  projectStartDate: string | null;
  /** 프로젝트 종료일 (ISO 8601) */
  projectEndDate: string | null;
  /** 프로젝트 범위/목적 */
  projectScope: string | null;
  /** 팀 참여 인원 수 */
  memberCount: number;
  /** 프로젝트 문서 총 수 */
  documentCount: number;
  /** 진행률 (0~100, 미집계 시 null) */
  progressRate: number | null;
}

/** 사업 개요 수정 요청 payload. 6개 필드를 명시적으로 전송한다. */
export interface UpdateBusinessOverviewPayload {
  /** 발주사명 */
  clientCompany: string | null;
  /** 수주사명 */
  contractorCompany: string | null;
  /** 계약 금액 (원화 단위 정수) */
  contractAmount: number | null;
  /** 프로젝트 시작일 (ISO 8601) */
  projectStartDate: string | null;
  /** 프로젝트 종료일 (ISO 8601) */
  projectEndDate: string | null;
  /** 프로젝트 범위/목적 */
  projectScope: string | null;
}
